Adds 'reset status' functionality

jensschuppe opened the pull request:
Fixes #383.
Replaces #392 (rebased onto current `master`) by @VangelisP, description at time of creation:

> This small addition implements a new functionality that will reset the status of the transaction from 'Processed' to 'New'. it will also remove the stored suggestions for that transaction.
> 
> This is being delivered into 2 parts:
> 
> * On the `api/v3/BankingTransaction.php` we add a new function called `civicrm_api3_banking_transaction_reset`. It accepts the transaction ID as the one and only parameter.
> * On the `/templates/CRM/Banking/Page/Review.tpl` we add a new button on the analysed and ignored' transactions, along with an ajax call to the new API command mentioned above.
> 
> Screenshots:
> 
> ![image](https://private-user-images.githubusercontent.com/14973113/251721838-a7612691-ad6b-43d4-9be5-c19aac611aed.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjU1NDA2NTAsIm5iZiI6MTc2NTU0MDM1MCwicGF0aCI6Ii8xNDk3MzExMy8yNTE3MjE4MzgtYTc2MTI2OTEtYWQ2Yi00M2Q0LTliZTUtYzE5YWFjNjExYWVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjEyVDExNTIzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2ZGQyMTg2ZDAyYTI3MWE4ZTJhN2FmMmE2ZDcwNGUwNmEzMDk2ZTU0NDAyODlhYTZmNTgxZjI0NTdkNzRmYmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.kkGY0ZO9WSro-s_nY-GVqlhF2sxcqyNbIQrmZTFxrEc)
> 
> ![image](https://private-user-images.githubusercontent.com/14973113/251722005-eebc46b4-1d9a-4e69-be78-53428bb3ac53.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjU1NDA2NTAsIm5iZiI6MTc2NTU0MDM1MCwicGF0aCI6Ii8xNDk3MzExMy8yNTE3MjIwMDUtZWViYzQ2YjQtMWQ5YS00ZTY5LWJlNzgtNTM0MjhiYjNhYzUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjEyVDExNTIzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkNTI1ZjBjOGZkYmJlZTRmNTZjMGM4ZjFmNjYxN2ZkMGVhYmJjMDJiODQ1MGJkYWYwNGEzMjI3YjYyMjIxYjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.qjaFArj8zrr6pac3MEl1Yrbt_qGlAbUjgGlMwBm_sUg)

jensschuppe committed:
Adds 'reset status' functionality. Implements #383

dontub commented on templates/CRM/Banking/Page/Review.tpl:
```diff
@@ -14,6 +14,17 @@
+  {assign var="resetLink" value=""}
+
+  {if $btxstatus.label eq 'Ignored'}
+    {if $allow_trx_reset}
```
Maybe it makes sense to combine it with the `if` before.

dontub commented on templates/CRM/Banking/Page/Review.tpl:
```diff
@@ -188,6 +202,57 @@ function analysePayment() {
+    return;
+  }
+
+  var reload_regex = new RegExp("(execute=)[0-9]+", 'ig');
```
The use of double and single quotes isn't consistent. (Here and in the rest of the code.)

dontub commented:
> This needs [systopia/civicrm-extension-template#4](https://github.com/systopia/civicrm-extension-template/pull/4) be implemented here for the PHPUnit action not to fail running, and it'll help to get the PHP_CodeSniffer and PHPStan actions come back green as well for merging this sooner.

I'm not sure what you mean with the second part of the sentence. Anyhow the phpcs and phpstan issues are independent of phpunit.

bjendres commented:
@jensschuppe do we still want this to be part of ``1.4``?

jensschuppe commented:
Probably not, I removed the milestone (we usually only add milestones when merging or if we really want things to be in the next version or for backports - not sure why I added it here).

gngn23 commented on CRM/Admin/Form/Setting/BankingSettings.php:
```diff
@@ -222,6 +230,9 @@ public function postProcess(): void {
     }
 
+    // allow trx status reset
+    Civi::settings()->set('allow_trx_reset', !empty($values['allow_trx_reset']));
```
At least for me
``` php
!empty($values['allow_trx_reset'])
```
is more readable than
``` php
(bool) ($values['allow_trx_reset'] ?? FALSE)
```

gngn23 commented on templates/CRM/Banking/Page/Review.tpl:
```diff
@@ -188,6 +202,57 @@ function analysePayment() {
+    return;
+  }
+
+  var reload_regex = new RegExp("(execute=)[0-9]+", 'ig');
```
I agree, we should use `const` (I think).
